internal/filecache/file_cache.go

```diff
@@ -23,55 +23,28 @@ func newFileCache(dir string) *fileCache {
 // Note: this can be expanded to do binary signing/verification, set TTL on each entry, etc.
 type fileCache struct {
 	dirPath string
-	mux     sync.RWMutex
-}
-
-type fileReadCloser struct {
-	*os.File
-	fc *fileCache
 }
 
 func (fc *fileCache) path(key Key) string {
 	return path.Join(fc.dirPath, hex.EncodeToString(key[:]))
 }
 
 func (fc *fileCache) Get(key Key) (content io.ReadCloser, ok bool, err error) {
-	// TODO: take lock per key for more efficiency vs the complexity of impl.
-	fc.mux.RLock()
-	unlock := fc.mux.RUnlock
-	defer func() {
-		if unlock != nil {
-			unlock()
-		}
-	}()
-
 	f, err := os.Open(fc.path(key))
 	if errors.Is(err, os.ErrNotExist) {
 		return nil, false, nil
 	} else if err != nil {
 		return nil, false, err
 	} else {
-		// Unlock is done inside the content.Close() at the call site.
-		unlock = nil
-		return &fileReadCloser{File: f, fc: fc}, true, nil
+		return f, true, nil
 	}
 }
 
-// Close wraps the os.File Close to release the read lock on fileCache.
-func (f *fileReadCloser) Close() (err error) {
-	defer f.fc.mux.RUnlock()
-	err = f.File.Close()
-	return
-}
-
 func (fc *fileCache) Add(key Key, content io.Reader) (err error) {
-	// TODO: take lock per key for more efficiency vs the complexity of impl.
-	fc.mux.Lock()
-	defer fc.mux.Unlock()
-
-	// Use rename for an atomic write
 	path := fc.path(key)
-	file, err := os.Create(path + ".tmp")
+	dirPath, fileName := filepath.Split(path)
+
+	file, err := os.CreateTemp(dirPath, fileName+".*.tmp")
```

Comment:
Can we do:

```go
	defer func() {
		if err != nil {
			_ = file.Close()
			_ = os.Remove(file.Name())
		}
	}()
```

And remove the `defer file.Close()`?
I'm not sure on Windows `os.Remove(…)` will work before `f.Close()` (and `defer`s run in reverse order).

Reply:
I made the change in [52d6eb1](https://github.com/tetratelabs/wazero/pull/2088/commits/52d6eb198af742efe98d196c0b9b7ace0935dc5a)

I made an adjustment to close the file regardless of whether an error occurred.